source/auth/auth.rst

```diff
@@ -1159,6 +1159,218 @@ If AWS authentication fails for any reason, the cache MUST be cleared.
+              description: "The OIDC refresh token"
+              type: str
+              optional: true
+
+.. code:: typescript
+
+    function onOIDCRequestToken(serverInfo: OIDCMechanismServerStep1): OIDCRequestTokenResult
+
+If the callback does not return an object in the correct form of ``OIDCRequestTokenResult``, the driver MUST raise an error.   The driver will
+inspect that the correct properties are given, but MUST NOT attempt to validate
+the token(s) directly.
+
+The optional token refresh callback will accept the IDP information as
+well as the cached OIDCRequestTokenResult and return a new OIDCRequestTokenResult.
+
+
+.. code:: typescript
+
+    function onOIDCRefreshToken(serverInfo: OIDCMechanismServerStep1, tokenResult: OIDCRequestTokenResult): OIDCRequestTokenResult
+
+If the callback does not return an object in the correct form of ``OIDCRequestTokenResult``, the driver MUST raise an error.
+
+If the refresh callback is given and the request callback is not given,
+the driver MUST raise an error.  If DEVICE_NAME is given and one or more
+callbacks are given, the driver MUST raise an error.
+
+If no callbacks are given, the driver MUST enforce that a DEVICE_NAME
+mechanism_properties is set and one of ("aws", "azure", or "gcp").
+The callback mechanism can be used to support both Authentication
+Code Workflows or Device workflows that are not explicitly implemented
+by drivers.  If there is no callback and no DEVICE_NAME, or the
+DEVICE_NAME is set but credentials cannot be automatically obtained,
+the driver MUST raise an error.
+
+
+Supported Device Workflows
+``````````````````````````
+
+Drives MUST support device workflows for "aws", "azure", and "gcp", given
+by the DEVICE_NAME mechanism property.  In all cases the acquired token
+will be given as the ``jwt`` argument in the second client step of the
+SASL exchange.
+
+AWS
+___
+
+When the DEVICE_NAME mechanism property is set to "aws", the driver MUST
+attempt to read the value given by the ``AWS_WEB_IDENTITY_TOKEN_FILE`` and
+interpret it as a file path.  The contents of the file are read as the
+access token.
+
+Azure
+_____
+
+When the DEVICE_NAME mechanism property is set to "azure", the driver MUST
+acquire an access token using the workflow described in "Obtaining an Access Token for Azure Key Vault" <TODO insert link to client-side-encryption>.
+
+GCP
+___
+
+When the DEVICE_NAME mechanism property is set to "gcp", the driver MUST
+acquire an access token using the workflow described in "Obtaining GCP Credentials" <TODO insert link to client-side-encryption>.
+
+
+Caching Credentials
+```````````````````
+
+Drivers MUST enable caching when callback(s) are provided to the mongo client.
+When an authorization request is made and there is a valid cached response,
+the driver MUST use the cached response if it has not expired.
+
+The cached responses are tied to the principal name if given, and an id of the callback function, if possible in the driver's language.
```

Comment:
Maybe I’m misunderstanding this, but:

I assume that this cache is implemented on a per-MongoClient basis. Would it be possible to change the callback function after a client has been instantiated? Otherwise I’m not quite sure what the purpose of including the id of the callback is.

Reply:
No, the cache is meant to be global, where the keys are some combination of the principal name and the callback hash.  We could clarify that if a per-client cache is used, then the only key needs to be the principal name.

Reply:
I’d add a clarifying note about that here, yes :slightly_smiling_face:

Reply:
Done!